### PR TITLE
Fix `euclidean_dist` in IVF-Flat search

### PR DESCRIPTION
**Pull request by Nyrio**
Solves #1058 

This is a tricky bug so the fix deserves some explanation. The previous implementation of `euclidean_dist` was the following in vectorized cases, where `x` and `y` are `int32` vectors of 4 `int8` each and `acc` is a single `int32` number to accumulate the distance in:

```c++
// Compute vectorized absolute differences independently.
const auto diff = static_cast<int32_t>(__vabsdiffs4(x, y));
// Square, reduce, and add to the accumulator.
acc = dp4a(diff, diff, acc);
```

Now consider the following case:

```c++
x = 0x80; // -128, 0, 0, 0
y = 0x7f; //  127, 0, 0, 0
```

The difference between -128 and 127 is 255, represented as `FF` (`__vabsdiffs4` is smart enough not to compute `abs(a-b)` which would result in `01`). However, if we call the signed version of `dp4a`, `FF` is cast from `int8` to `int32` as `FFFFFFFF` (or -1). The square of -1 is 1, which is added to `acc` (instead of 65025).

As the output of `__vabsdiffs4` is correct when considered as an unsigned number, and as addition is the same for signed and unsigned in 2's complement (and `acc` is positive anyway), the easiest fix is to use the unsigned version of `dp4a`, which will cast overflowed differences properly to 32 bits. The previous code simply becomes:

```c++
const auto diff = __vabsdiffs4(x, y);
acc = dp4a(diff, diff, static_cast<uint32_t>(acc));
```

-----

Additionally, to avoid underflows in the non-vectorized unsigned case, I replaced the subtraction with `__usad` (absolute difference of unsigned numbers). Note that using the subtraction was correct anyway, because the addition/subtraction is the same for unsigned and signed integers, as well as the least significant half of the multiplication (which is the part that is stored), and the square of a number is also the square of its opposite. Consider:

```c++
uint32_t a = 10;
uint32_t b = 20;
uint32_t c = a - b; // fffffff6, i.e -10 or 4294967286
uint32_t d = c * c; // (ffffffec)00000064, i.e 100
```

### CODE REVIEW

File: cpp/include/raft/spatial/knn/detail/ivf_flat_search.cuh [hunk 1]
```diff
@@ -897,8 +898,8 @@ struct euclidean_dist<Veclen, int8_t, int32_t> {
   __device__ __forceinline__ void operator()(int32_t& acc, int32_t x, int32_t y)
   {
     if constexpr (Veclen > 1) {
-      const auto diff = static_cast<int32_t>(__vabsdiffs4(x, y));
-      acc             = dp4a(diff, diff, acc);
+      const auto diff = __vabsdiffs4(x, y);
+      acc             = dp4a(diff, diff, static_cast<uint32_t>(acc));
```
Comment (achirkin) [inline cpp/include/raft/spatial/knn/detail/ivf_flat_search.cuh:902]: So, do I understand the problem right, that `dp4a.s32.s32` interprets each byte as `int8_t`, whereas `__vabsdiffs4` produces four `uint8_t` values (thus, the outputs above 127 interpreted as negative)? And the solution is to do `dp4a.u32.u32`, which interprets components correctly as `uint8_t`? Then, could you please maybe add a short note here that we really want the unsigned overload of `dp4a`?

File: cpp/include/raft/spatial/knn/detail/ivf_flat_search.cuh [hunk 2]
```diff
@@ -886,6 +886,7 @@ struct euclidean_dist<Veclen, uint8_t, uint32_t> {
       const auto diff = __vabsdiffu4(x, y);
       acc             = dp4a(diff, diff, acc);
     } else {
+      // Note: diff can underflow but the multiplication is still correct in that case.
```
Comment (achirkin) [inline cpp/include/raft/spatial/knn/detail/ivf_flat_search.cuh:889]: Now that you said it out loud, it feels a little uncomfortable. Maybe use `x > y ? x - y : y - x` or `__usad ( unsigned int  x, unsigned int  y, unsigned int  z)`?

**Review by achirkin — CHANGES_REQUESTED** [review]
Wow, that must have been a tough investigation, good job! You've explained the problem very nice here, but maybe it's worth adding a short gist of it (or a link) in the code?

File: cpp/include/raft/spatial/knn/detail/ivf_flat_search.cuh (hunk 2 above)
Comment (Nyrio) [reply cpp/include/raft/spatial/knn/detail/ivf_flat_search.cuh:889]: `x > y ? x - y : y - x` is less efficient than casting to signed.

I could go with `__usad`, it switches one instruction for another: `IADD3` to `VABSDIFF.U32`.

**Review by tfeher — APPROVED** [review]
Thanks Louis for fixing this problem, the PR looks good to me!

Could you update the second part of the PR description? You say
> added comment in the unsigned, non-vectorized implementation

But that has changed during the review.

**Comment by Nyrio** [comment]
> Could you update the second part of the PR description?

Ah, yes, done.

**Review by cjnolet — APPROVED** [review]
Changes LGTM. Thanks @Nyrio!

**Review by achirkin — APPROVED** [review]
LGTM!

**Pull request merged**